File: flyweight/java/Restaurant.java
package flyweight.java;

import java.util.List;

public class Restaurant {
    private List<Food> foods = new java.util.ArrayList<Food>();

    public void addFood(int cost, String name, String cuisine, String category) {
        FoodType foodType = FoodFactory.getFoodType(cuisine, category);
        Food food = new Food(cost, name, foodType);
        foods.add(food);
    }

    public void serveFood() {
        for (Food food : foods) {
            food.consume();
        }
    }
}

File: flyweight/golang/restaurant.go
package main

type Restaurant struct {
	foods []*Food
}

func (r *Restaurant) AddFood(cost float64, name string,
	cusine string, category string) {

	foodFactory := NewFoodFactory()
	foodType := foodFactory.GetFoodType(cusine, category)
	food := &Food{
		price:    cost,
		name:     name,
		foodType: foodType,
	}

	r.foods = append(r.foods, food)
}

func (r *Restaurant) ServeFood() {
	for _, food := range r.foods {
		println(food.Serve())
	}
}
